Grab eComment link mapping from meetings.js

hancush opened the pull request:
### Description

eComment links are rendered by a JavaScript event, which is a pain for us, because we can't scrape them.

This PR spikes one method of mapping events to their eComment links by creating a lookup table from the JavaScript file located at [`https://metro.granicusideas.com/meetings.js`](https://metro.granicusideas.com/meetings.js).

This file is loaded on both the [event listing page](https://metro.legistar.com/Calendar.aspx), and [event detail](https://metro.legistar.com/MeetingDetail.aspx?LEGID=1718&GID=557&G=A5FAA737-A54D-4A6C-B1E8-FF70F765FA94) pages. I do not see any evidence on either page that it's filterable, e.g., shows only eComments links for visible events on a listing page. I have also confirmed that, when multiple eComment pages are enabled, they all show up in this JavaScript file.

### Alternatives

eComment links look like `https://metro.granicusideas.com/meetings/750-finance-budget-and-audit-committee-on-2020-03-16-5-00-pm-test`. One alternative I considered is proposing a standard format for eComment titles (if they are in fact editable by Metro) and programmatically building the slug for the URL.

It looks like the meeting ID is automatically prepended (?), so we'd still need to parse that from the detail table, but the slug generation should happen on the event scraper, instead of on the base class, since it requires event details.

I decided to prefer the JavaScript approach, because relying on manual data entry in the past has caused us some big headaches. If the proposed approach does not work as I think it does, then this could be a feasible fallback.

### Notes

It would be better code organization to define these changes on the event scraper, but it also feels coupled tightly enough to detail table parsing that it can live on the base class. I [spiked moving it to the event scraper](https://github.com/opencivicdata/python-legistar-scraper/compare/feature/hec/scrape-ecomment-link...spike/hec/move-comments-to-event), but we lose some of the niceness of our API event scraping separation. Happy for your feedback here.

fgregg commented on legistar/base.py:
```diff
@@ -79,6 +79,32 @@ class LegistarScraper(scrapelib.Scraper, LegistarSession):
         super(LegistarScraper, self).__init__(*args, **kwargs)
 
+    @property
+    def ecomment_dict(self):
```
this seems like something that a regex would be good for, but this seems fine for now.

fgregg replied:
https://github.com/Kronuz/esprima-python could be a good alternative.

fgregg commented on legistar/base.py:
```diff
@@ -139,6 +165,8 @@ def parseDetails(self, detail_div):
                 for link in field_2.xpath('.//a'):
                     value.append({'label': link.text_content().strip(),
                                   'url': self._get_link_address(link)})
+            elif key == 'eComment':
```
a way that would be pretty clean to factor this out to the event class is to have this block look like 

```python
            if field_2.find('.//a') is not None:
                value = []
                for link in field_2.xpath('.//a'):
                    value.append({'label': link.text_content().strip(),
                                  'url': self._get_link_address(link)})
            elif 'href' in field_2.attrib:
                value = {'label': field_2.text_content().strip(),
                         'url': self._get_link_address(field_2)}
            elif self._detail_parse(field):
                value = self._detail_parse(field)
            else:
                value = field_2.text_content().strip()
```

and define `_detail_parse` to return None on the base class, and override on the event class.

fgregg commented on legistar/base.py:
```diff
@@ -79,6 +79,32 @@ class LegistarScraper(scrapelib.Scraper, LegistarSession):
+        if not getattr(self, '_ecomment_dict', None):
+            ecomment_dict = {}
+
+            script = requests.get('https://metro.granicusideas.com/meetings.js')
```
the URL needs to be paramaterized, i.e. self.ECOMMENT_JS_URL

fgregg commented:
overall, seems like a sane approach!

hancush commented on legistar/base.py:
```diff
@@ -127,21 +127,28 @@ def parseDetails(self, detail_div):
                        "     or starts-with(@id, 'ctl00_ContentPlaceHolder1_hyp')"\
                        "     or starts-with(@id, 'ctl00_ContentPlaceHolder1_Label')]"
         fields = detail_div.xpath(detail_query)
+
```
Added some whitespace here for legibility.

hancush commented on legistar/events.py:
```diff
@@ -11,6 +12,49 @@
+            response = self.get(self.ECOMMENT_JS_URL)
+            tree = esprima.parseScript(response.text)
+
+            try:
```
Both the exception handling and traversal of the parsed script feel pretty clunky! Any suggestions for improvement are welcome. Here's what the parsed script looks like:

```python
{
    type: "Program",
    sourceType: "script",
    body: [
        {
            type: "FunctionDeclaration",
            expression: False,
            isAsync: False,
            id: {
                type: "Identifier",
                name: "activateEcomment"
            },
            params: [
                {
                    type: "Identifier",
                    name: "event_id"
                },
                {
                    type: "Identifier",
                    name: "event_uid"
                },
                {
                    type: "Identifier",
                    name: "url"
                }
            ],
            body: {
                type: "BlockStatement",
                body: [
                    {
                        type: "VariableDeclaration",
                        declarations: [
                            {
                                type: "VariableDeclarator",
                                id: {
                                    type: "Identifier",
                                    name: "ele"
                                },
                                init: {
                                    type: "CallExpression",
                                    callee: {
                                        type: "Identifier",
                                        name: "gQuery"
                                    },
                                    arguments: [
                                        {
                                            type: "BinaryExpression",
                                            operator: "+",
                                            left: {
                                                type: "BinaryExpression",
                                                operator: "+",
                                                left: {
                                                    type: "BinaryExpression",
                                                    operator: "+",
                                                    left: {
                                                        type: "BinaryExpression",
                                                        operator: "+",
                                                        left: {
                                                            type: "BinaryExpression",
                                                            operator: "+",
                                                            left: {
                                                                type: "BinaryExpression",
                                                                operator: "+",
                                                                left: {
                                                                    type: "Literal",
                                                                    value: "[rel='ecomment'][data-event-id='",
                                                                    raw: "\"[rel='ecomment'][data-event-id='\""
                                                                },
                                                                right: {
                                                                    type: "Identifier",
                                                                    name: "event_id"
                                                                }
                                                            },
                                                            right: {
                                                                type: "Literal",
                                                                value: "'],[rel='ecomment'][data-event-id='",
                                                                raw: "\"'],[rel='ecomment'][data-event-id='\""
                                                            }
                                                        },
                                                        right: {
                                                            type: "CallExpression",
                                                            callee: {
                                                                type: "MemberExpression",
                                                                computed: False,
                                                                object: {
                                                                    type: "Identifier",
                                                                    name: "event_uid"
                                                                },
                                                                property: {
                                                                    type: "Identifier",
                                                                    name: "toUpperCase"
                                                                }
                                                            },
                                                            arguments: []
                                                        }
                                                    },
                                                    right: {
                                                        type: "Literal",
                                                        value: "'],[rel='ecomment'][data-event-id='",
                                                        raw: "\"'],[rel='ecomment'][data-event-id='\""
                                                    }
                                                },
                                                right: {
                                                    type: "CallExpression",
                                                    callee: {
                                                        type: "MemberExpression",
                                                        computed: False,
                                                        object: {
                                                            type: "Identifier",
                                                            name: "event_uid"
                                                        },
                                                        property: {
                                                            type: "Identifier",
                                                            name: "toLowerCase"
                                                        }
                                                    },
                                                    arguments: []
                                                }
                                            },
                                            right: {
                                                type: "Literal",
                                                value: "']",
                                                raw: "\"']\""
                                            }
                                        }
                                    ]
                                }
                            }
                        ],
                        kind: "var"
                    },
                    {
                        type: "ExpressionStatement",
                        expression: {
                            type: "CallExpression",
                            callee: {
                                type: "MemberExpression",
                                computed: False,
                                object: {
                                    type: "CallExpression",
                                    callee: {
                                        type: "MemberExpression",
                                        computed: False,
                                        object: {
                                            type: "Identifier",
                                            name: "ele"
                                        },
                                        property: {
                                            type: "Identifier",
                                            name: "attr"
                                        }
                                    },
                                    arguments: [
                                        {
                                            type: "Literal",
                                            value: "href",
                                            raw: "\"href\""
                                        },
                                        {
                                            type: "Identifier",
                                            name: "url"
                                        }
                                    ]
                                },
                                property: {
                                    type: "Identifier",
                                    name: "show"
                                }
                            },
                            arguments: []
                        }
                    },
                    {
                        type: "VariableDeclaration",
                        declarations: [
                            {
                                type: "VariableDeclarator",
                                id: {
                                    type: "Identifier",
                                    name: "insiteLink"
                                },
                                init: {
                                    type: "CallExpression",
                                    callee: {
                                        type: "Identifier",
                                        name: "gQuery"
                                    },
                                    arguments: [
                                        {
                                            type: "BinaryExpression",
                                            operator: "+",
                                            left: {
                                                type: "BinaryExpression",
                                                operator: "+",
                                                left: {
                                                    type: "BinaryExpression",
                                                    operator: "+",
                                                    left: {
                                                        type: "BinaryExpression",
                                                        operator: "+",
                                                        left: {
                                                            type: "BinaryExpression",
                                                            operator: "+",
                                                            left: {
                                                                type: "BinaryExpression",
                                                                operator: "+",
                                                                left: {
                                                                    type: "Literal",
                                                                    value: "form[name='aspnetForm'] .ecomment[data-event-id='",
                                                                    raw: "\"form[name='aspnetForm'] .ecomment[data-event-id='\""
                                                                },
                                                                right: {
                                                                    type: "Identifier",
                                                                    name: "event_id"
                                                                }
                                                            },
                                                            right: {
                                                                type: "Literal",
                                                                value: "'],form[name='aspnetForm'] .ecomment[data-event-id='",
                                                                raw: "\"'],form[name='aspnetForm'] .ecomment[data-event-id='\""
                                                            }
                                                        },
                                                        right: {
                                                            type: "CallExpression",
                                                            callee: {
                                                                type: "MemberExpression",
                                                                computed: False,
                                                                object: {
                                                                    type: "Identifier",
                                                                    name: "event_uid"
                                                                },
                                                                property: {
                                                                    type: "Identifier",
                                                                    name: "toLowerCase"
                                                                }
                                                            },
                                                            arguments: []
                                                        }
                                                    },
                                                    right: {
                                                        type: "Literal",
                                                        value: "'],form[name='aspnetForm'] .ecomment[data-event-id='",
                                                        raw: "\"'],form[name='aspnetForm'] .ecomment[data-event-id='\""
                                                    }
                                                },
                                                right: {
                                                    type: "CallExpression",
                                                    callee: {
                                                        type: "MemberExpression",
                                                        computed: False,
                                                        object: {
                                                            type: "Identifier",
                                                            name: "event_uid"
                                                        },
                                                        property: {
                                                            type: "Identifier",
                                                            name: "toUpperCase"
                                                        }
                                                    },
                                                    arguments: []
                                                }
                                            },
                                            right: {
                                                type: "Literal",
                                                value: "']",
                                                raw: "\"']\""
                                            }
                                        }
                                    ]
                                }
                            }
                        ],
                        kind: "var"
                    },
                    {
                        type: "ExpressionStatement",
                        expression: {
                            type: "CallExpression",
                            callee: {
                                type: "MemberExpression",
                                computed: False,
                                object: {
                                    type: "CallExpression",
                                    callee: {
                                        type: "MemberExpression",
                                        computed: False,
                                        object: {
                                            type: "CallExpression",
                                            callee: {
                                                type: "MemberExpression",
                                                computed: False,
                                                object: {
                                                    type: "CallExpression",
                                                    callee: {
                                                        type: "MemberExpression",
                                                        computed: False,
                                                        object: {
                                                            type: "CallExpression",
                                                            callee: {
                                                                type: "MemberExpression",
                                                                computed: False,
                                                                object: {
                                                                    type: "CallExpression",
                                                                    callee: {
                                                                        type: "MemberExpression",
                                                                        computed: False,
                                                                        object: {
                                                                            type: "CallExpression",
                                                                            callee: {
                                                                                type: "MemberExpression",
                                                                                computed: False,
                                                                                object: {
                                                                                    type: "Identifier",
                                                                                    name: "insiteLink"
                                                                                },
                                                                                property: {
                                                                                    type: "Identifier",
                                                                                    name: "text"
                                                                                }
                                                                            },
                                                                            arguments: [
                                                                                {
                                                                                    type: "Literal",
                                                                                    value: "eComment",
                                                                                    raw: "\"eComment\""
                                                                                }
                                                                            ]
                                                                        },
                                                                        property: {
                                                                            type: "Identifier",
                                                                            name: "attr"
                                                                        }
                                                                    },
                                                                    arguments: [
                                                                        {
                                                                            type: "Literal",
                                                                            value: "href",
                                                                            raw: "\"href\""
                                                                        },
                                                                        {
                                                                            type: "Identifier",
                                                                            name: "url"
                                                                        }
                                                                    ]
                                                                },
                                                                property: {
                                                                    type: "Identifier",
                                                                    name: "attr"
                                                                }
                                                            },
                                                            arguments: [
                                                                {
                                                                    type: "Literal",
                                                                    value: "target",
                                                                    raw: "\"target\""
                                                                },
                                                                {
                                                                    type: "Literal",
                                                                    value: "_blank",
                                                                    raw: "\"_blank\""
                                                                }
                                                            ]
                                                        },
                                                        property: {
                                                            type: "Identifier",
                                                            name: "css"
                                                        }
                                                    },
                                                    arguments: [
                                                        {
                                                            type: "Literal",
                                                            value: "color",
                                                            raw: "\"color\""
                                                        },
                                                        {
                                                            type: "Literal",
                                                            value: "blue",
                                                            raw: "\"blue\""
                                                        }
                                                    ]
                                                },
                                                property: {
                                                    type: "Identifier",
                                                    name: "css"
                                                }
                                            },
                                            arguments: [
                                                {
                                                    type: "Literal",
                                                    value: "text-decoration",
                                                    raw: "\"text-decoration\""
                                                },
                                                {
                                                    type: "Literal",
                                                    value: "underline",
                                                    raw: "\"underline\""
                                                }
                                            ]
                                        },
                                        property: {
                                            type: "Identifier",
                                            name: "css"
                                        }
                                    },
                                    arguments: [
                                        {
                                            type: "Literal",
                                            value: "cursor",
                                            raw: "\"cursor\""
                                        },
                                        {
                                            type: "Literal",
                                            value: "pointer",
                                            raw: "\"pointer\""
                                        }
                                    ]
                                },
                                property: {
                                    type: "Identifier",
                                    name: "addClass"
                                }
                            },
                            arguments: [
                                {
                                    type: "Literal",
                                    value: "active",
                                    raw: "\"active\""
                                }
                            ]
                        }
                    }
                ]
            },
            generator: False
        },
        {
            type: "ExpressionStatement",
            expression: {
                type: "CallExpression",
                callee: {
                    type: "Identifier",
                    name: "gQuery"
                },
                arguments: [
                    {
                        type: "FunctionExpression",
                        expression: False,
                        isAsync: False,
                        params: [],
                        body: {
                            type: "BlockStatement",
                            body: [
                                {
                                    type: "ExpressionStatement",
                                    expression: {
                                        type: "CallExpression",
                                        callee: {
                                            type: "Identifier",
                                            name: "activateEcomment"
                                        },
                                        arguments: [
                                            {
                                                type: "Literal",
                                                value: "752",
                                                raw: "'752'"
                                            },
                                            {
                                                type: "Literal",
                                                value: "8ADA023E-1316-4675-8479-607EC2FC8E73",
                                                raw: "'8ADA023E-1316-4675-8479-607EC2FC8E73'"
                                            },
                                            {
                                                type: "Literal",
                                                value: "https://metro.granicusideas.com/meetings/752-finance-budget-and-audit-committee-on-2020-03-17-11-00-am-test",
                                                raw: "'https://metro.granicusideas.com/meetings/752-finance-budget-and-audit-committee-on-2020-03-17-11-00-am-test'"
                                            }
                                        ]
                                    }
                                },
                                {
                                    type: "ExpressionStatement",
                                    expression: {
                                        type: "CallExpression",
                                        callee: {
                                            type: "Identifier",
                                            name: "activateEcomment"
                                        },
                                        arguments: [
                                            {
                                                type: "Literal",
                                                value: "751",
                                                raw: "'751'"
                                            },
                                            {
                                                type: "Literal",
                                                value: "E6DC5C58-D682-4B53-B78D-35E3B70AC7CA",
                                                raw: "'E6DC5C58-D682-4B53-B78D-35E3B70AC7CA'"
                                            },
                                            {
                                                type: "Literal",
                                                value: "https://metro.granicusideas.com/meetings/751-finance-budget-and-audit-committee-on-2020-03-17-12-00-pm-test",
                                                raw: "'https://metro.granicusideas.com/meetings/751-finance-budget-and-audit-committee-on-2020-03-17-12-00-pm-test'"
                                            }
                                        ]
                                    }
                                }
                            ]
                        },
                        generator: False
                    }
                ]
            }
        }
    ]
}
```

fgregg replied:
hi @hancush, good job!

Here's a cleaner way of doing this.

```
import esprima

js = """function activateEcomment(event_id, event_uid, url) {
  // For MediaManager
  var ele = gQuery("[rel='ecomment'][data-event-id='" + event_id + "'],[rel='ecomment'][data-event-id='" + event_uid.toUpperCase() + "'],[rel='ecomment'][data-event-id='" + event_uid.toLowerCase() + "']");
  ele.attr("href", url).show();

  // For Insite
  var insiteLink = gQuery("form[name='aspnetForm'] .ecomment[data-event-id='" + event_id + "'],form[name='aspnetForm'] .ecomment[data-event-id='" + event_uid.toLowerCase() + "'],form[name='aspnetForm'] .ecomment[data-event-id='" + event_uid.toUpperCase() + "']");

  insiteLink
    .text("eComment")
    .attr("href", url)
    .attr("target", "_blank")
    .css("color", "blue")
    .css("text-decoration", "underline")
    .css("cursor", "pointer")
    .addClass("active");
}

  gQuery(function () {
    activateEcomment('751', 'E6DC5C58-D682-4B53-B78D-35E3B70AC7CA', 'https://metro.granicusideas.com/meetings/751-finance-budget-and-audit-committee-on-2020-03-17-12-00-pm-test');
  });
"""


def activateEcomment_args(js_text):
    results = []

    def is_activateComment(node, metadata):
        if node.callee and node.callee.name == 'activateEcomment':
            results.append([arg.value for arg in node.arguments])

    esprima.parse(js_text, delegate=is_activateComment)

    return results

print(activateEcomment_args(js))
```

fgregg replied:
here's a docs for what I did here. https://esprima.readthedocs.io/en/latest/syntactic-analysis.html#example-console-calls-removal

hancush commented:
@fgregg Took a shot at addressing your comments, because I'm not sure when we'd be able to circle back, and JavaScript parsing will (hopefully) be more reliable than what I was doing before. Left some comments inline where specific feedback would be useful.

hancush commented:
FYI: I'm going to merge this so we can get the changes live, but I would still value review!
